cpp/examples/arrow/execution_plan_documentation_examples.cc

```diff
@@ -0,0 +1,1160 @@
+    // // // start the ExecPlan
+    ABORT_ON_FAILURE(plan->StartProducing());
+    plan->finished().Wait();
+    return arrow::Status::OK();
+}
+
+arrow::Status SourceUnionSinkExample() {
+    auto basic_data = MakeBasicBatches();
+
+    cp::ExecContext exec_context(arrow::default_memory_pool(),
+                                 ::arrow::internal::GetCpuThreadPool());
+
+    std::shared_ptr<cp::ExecPlan> plan =
+    cp::ExecPlan::Make(&exec_context).ValueOrDie();
+    arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> sink_gen;
+
+    cp::Declaration union_node{"union", cp::ExecNodeOptions{}};
+    cp::Declaration lhs{"source",
+                  cp::SourceNodeOptions{basic_data.schema,
+                                    basic_data.gen(/*parallel=*/false)}};
+    lhs.label = "lhs";
+    cp::Declaration rhs{"source",
+                    cp::SourceNodeOptions{basic_data.schema,
+                                      basic_data.gen(/*parallel=*/false)}};
+    rhs.label = "rhs";
+    union_node.inputs.emplace_back(lhs);
+    union_node.inputs.emplace_back(rhs);
+
+    cp::CountOptions options(cp::CountOptions::ONLY_VALID);
+    ARROW_ASSIGN_OR_RAISE(auto declr,
+    cp::Declaration::Sequence(
+            {
+                union_node,
+                {"aggregate", cp::AggregateNodeOptions{
+                  /*aggregates=*/{{"count", &options}},
+                  /*targets=*/{"a"},
+                  /*names=*/{"count(a)"},
+                  /*keys=*/{}}},
+                {"sink", cp::SinkNodeOptions{&sink_gen}},
+            })
+            .AddToPlan(plan.get()));
+
+    ABORT_ON_FAILURE(declr->Validate());
+
+    ABORT_ON_FAILURE(plan->Validate());
+    std::shared_ptr<arrow::RecordBatchReader> sink_reader = cp::MakeGeneratorReader(
+        arrow::schema({arrow::field("count(a)", arrow::int32())}),
+        std::move(sink_gen),
+        exec_context.memory_pool());
+
+    // // // start the ExecPlan
+    ABORT_ON_FAILURE(plan->StartProducing());
+
+    // // collect sink_reader into a Table
+    std::shared_ptr<arrow::Table> response_table;
+
+    ARROW_ASSIGN_OR_RAISE(response_table,
+    arrow::Table::FromRecordBatchReader(sink_reader.get()));
+
+    std::cout << "Results : " << response_table->ToString() << std::endl;
+    return arrow::Status::OK();
+}
+
+enum ExampleMode {
+  SOURCE_SINK = 0,
+  SCAN_SINK = 1,
+  SCAN_FILTER_SINK = 2,
+  SCAN_PROJECT_SINK = 3,
+  SOURCE_AGGREGATE_SINK = 4,
+  SCAN_CONSUMING_SINK = 5,
+  SCAN_ORDER_BY_SINK = 6,
+  SCAN_HASHJOIN_SINK = 7,
+  SCAN_SELECT_SINK = 8,
+  SCAN_FILTER_WRITE = 9,
+  SOURCE_UNION_SINK = 10,
+};
+
+int main(int argc, char** argv) {
+  if (argc < 2) {
+    // Fake success for CI purposes.
+    return EXIT_SUCCESS;
+  }
+
+  int mode = std::atoi(argv[1]);
+  std::string base_save_path = argv[2];
+
+  switch (mode) {
+    case SOURCE_SINK:
+      PRINT_BLOCK("Source Sink Example");
+      CHECK_AND_CONTINUE(SourceSinkExample())
```

Comment:
We can mostly avoid defining new macros if we instead refactor this into a function that returns Status, and then just check the status once in `main`.